### DISCUSSION TITLE
I Have Abused Floor Plan (this is a good thing)

### DISCUSSION THREAD

**Discussion by tchoward** [ExperienceLovelace/ha-floorplan#149] [Ideas]
First of all, this is an excellent (extension).  I want to thank the creators and maintainers of this Lovelace Frontend.  I use it WAY too much.  Below are the floor plans I use on my phone to control my devices and get status.

<img width="382" alt="Screen Shot 2021-09-24 at 6 07 32 PM" src="https://user-images.githubusercontent.com/64616204/134744559-d11416b8-1dea-4834-9ac7-813bead8592e.png">
<img width="632" alt="Screen Shot 2021-09-24 at 6 08 41 PM" src="https://user-images.githubusercontent.com/64616204/134744648-5cc204b8-59a0-4e19-ac3a-063e4d853f01.png">

Here's the thing.  I have extended it's use to not only do floor plans, but also weather and solar energy status.  The screenshot below is the home panel I use for my bedroom and kitchen.  The wife loves it.  Everything EXCEPT the graph and radar are done EXCLUSIVELY in Floor Plan.  

<img width="1284" alt="Screen Shot 2021-09-24 at 6 12 05 PM" src="https://user-images.githubusercontent.com/64616204/134744889-808509c3-5bb6-49d9-8ec1-6de4b13aeca0.png">

So here is my question; has anyone given thought to extending the service capability?  For example, css-set or svg-set?  I found ways around most of it, but as an example, the balls in the power-flow card change animation time based on power usage.  To accomplish this I had to add a function like so:

```
setSpeed_: (element, speed) => {

        var animate = element.getElementsByTagName('animateMotion').item(0);
        speed = Math.abs(speed);

        var how_fast = 1;
        if (speed < 6){
          how_fast = 6 - speed;
        }

        animate.setAttribute("dur", how_fast.toFixed(1)+"s");  
        return 0;
        
      },

```
It works, but feels a bit...hacky.  

Thoughts?

**Comment by OzGav** [comment]
Firstly, awesome work!

I am not sure what you need css-set and svg-set to do that isn’t already available with class_set (or style_set) and image_set?

**Reply by tchoward** [reply]
Good evening, thanks for the response.  I guess the difficulty I am having (and maybe there is a way to do this; if so, I apologize) is being able to dynamically set values.  As a different example; I want to fill a battery.  I can do this by defining an svg rectangle.  Ideally, I could just modify the rectangle's width depending on the battery charge's value.  Today, I have to define 10 svg images which I use image-set.  I would prefer more granularity. Make sense?

I don't want to stir up anything, and I know I am stretching the original intent, so if it is not the cards... no issue.  Over on Hubitat I implemented HubiGraph which turned into a "feature request nightmare".  Don't want to do that here...

**Reply by OzGav** [reply]
Yes that might be out of scope and would need to @pkozul to comment but to date there has been no mention of manipulating the underlying SVG that I am aware of. In regards to your battery example I believe there is a solution to that in the Home example. Have a look at the rectangle that is the moisture sensor. https://experiencelovelace.github.io/ha-floorplan/docs/example-home/

**Reply by exetico** [reply]
Hi! 

That's impressive, and kinda hacky - as you're saying, too.

Don't you think the "Home" example includes what you need, for adjusting the height of an element? You're able to do many things with CSS - and Floorplan are allowing you to do, almost whatever you want.

Do you have other needs, which are not already mentioned here?

[ha-floorplan has a lib-structure](https://github.com/ExperienceLovelace/ha-floorplan/blob/master/src/components/floorplan/lib/color-util.ts), so it's easy to add helper-functions, if you're searching for a way to make some of the things a bit easier to navigate within? (Your YAML-code). We're happy to add other helper-functions too. Create a pull-request, if you have something ready for the main-repo.
